cluster/etcd_service_discovery.go

```diff
@@ -230,10 +236,11 @@ func (sd *etcdServiceDiscovery) deleteServer(serverID string) {
 	if actual, ok := sd.serverMapByID.Load(serverID); ok {
 		sv := actual.(*Server)
 		sd.serverMapByID.Delete(sv.ID)
-		if svMap, ok := sd.serverMapByType.Load(sv.Type); ok {
-			sm := svMap.(map[string]*Server)
-			delete(sm, sv.ID)
+		sd.mapByTypeLock.Lock()
```

Comment:
Isn't it better to `defer sd.mapByTypeLock.Unlock()` right after locking?

Reply:
Given that there is only one flow deferring is slower and would be executed only after notifying the listeners

Reply:
Ok, I created a wrapper for the lock, what do you think?